### PR TITLE
Support transforming params of arrow functions in class fields

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes https://github.com/babel/babel/issues/13939
| Patch: Bug Fix?          | Yes
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/13941"></a>

### CODE REVIEW

File: packages/babel-plugin-transform-parameters/test/fixtures/assumption-noNewArrows-false/default/output.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+var _this = this;
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-parameters/test/fixtures/assumption-noNewArrows-false/default/output.js:1]: This is another bug fixed by the new `if (!path.isFunction()) return;` check in the plugin.

File: packages/babel-plugin-transform-parameters/src/index.js [hunk 2]
```diff
@@ -23,6 +23,10 @@ export default declare((api, options) => {
         ) {
           // default/rest visitors require access to `arguments`, so it cannot be an arrow
           path.arrowFunctionToExpression({ noNewArrows });
+
+          // In some cases arrowFunctionToExpression replaces the function with a wrapper.
+          // Return early; the wrapped function will be visited later in the AST traversal.
+          if (!path.isFunction()) return;
```
Comment (JLHwung) [inline packages/babel-plugin-transform-parameters/src/index.js:29]: Q: Can this check be strengthen to `path.isFunctionExpression()`?

File: packages/babel-traverse/src/path/conversion.ts [hunk 3]
```diff
@@ -195,24 +195,44 @@ export function arrowFunctionToExpression(
...
  */
 function hoistFunctionEnvironment(
-  fnPath,
+  fnPath: NodePath<t.Function>,
   // TODO(Babel 8): Consider defaulting to `false` for spec compliancy
   noNewArrows = true,
   allowInsertArrow = true,
-) {
-  const thisEnvFn = fnPath.findParent(p => {
+): { thisBinding: string; fnPath: NodePath<t.Function> } {
```
Comment (JLHwung) [inline packages/babel-traverse/src/path/conversion.ts:202]: Can you add some comments about the return type? Especially we may have different `fnPath` output than the input.

File: packages/babel-traverse/src/path/conversion.ts [hunk 4]
```diff
@@ -195,24 +195,44 @@ export function arrowFunctionToExpression(
...
-      (p.isFunction() && !p.isArrowFunctionExpression()) ||
-      p.isProgram() ||
-      p.isClassProperty({ static: false })
+      p.isFunction() || p.isProgram() || p.isClassProperty({ static: false })
     );
   });
-  const inConstructor = thisEnvFn?.node.kind === "constructor";
+  const inConstructor = thisEnvFn.isClassMethod({ kind: "constructor" });
 
   if (thisEnvFn.isClassProperty()) {
```
Comment (JLHwung) [inline packages/babel-traverse/src/path/conversion.ts:215]:
```suggestion
  if (thisEnvFn.isClassProperty() || thisEnvFn.isClassPrivateProperty()) {
```

**Review by JLHwung — APPROVED** [review]

File: packages/babel-plugin-transform-parameters/src/index.js (hunk 2 above)
Comment (nicolo-ribaudo) [reply packages/babel-plugin-transform-parameters/src/index.js:29]: Yes, currently `path` is guaranteed to be either a `FunctionExpression` or a `CallExpression`

**Review by existentialism — APPROVED** [review]

**Pull request merged**